src/SortingAlgorithms.jl

```diff
@@ -9,16 +9,42 @@ using Base.Order
 import Base.Sort: sort!
 import DataStructures: heapify!, percolate_down!
 
-export HeapSort, TimSort, RadixSort
+export HeapSort, TimSort, RadixSort, CombSort
 
 struct HeapSortAlg  <: Algorithm end
 struct TimSortAlg   <: Algorithm end
 struct RadixSortAlg <: Algorithm end
+struct CombSortAlg  <: Algorithm end
 
 const HeapSort  = HeapSortAlg()
 const TimSort   = TimSortAlg()
 const RadixSort = RadixSortAlg()
 
+"""
+    CombSort
+
+Indicates that a sorting function should use the comb sort
+algorithm. Comb sort traverses the collection multiple times
+ordering pairs of elements with a given interval between them.
+The interval decreases exponentially until it becomes 1, then
+it switches to insertion sort on the whole input.
+
+Characteristics:
+ - *not stable* does not preserve the ordering of elements which
+   compare equal (e.g. "a" and "A" in a sort of letters which
+   ignores case).
+ - *in-place* in memory.
+ - *parallelizable* suitable for vectorization with SIMD instructions
+   because it performs many independent comparisons.
+ - *complexity* worst-case only proven to be better than quadratic, but not `n*log(n)`.
```

Comment:
This algorithm has quadratic worst-case runtime.

```julia-repl
julia> @time sort!(4^7*repeat(1:30, 4^7));
  0.027213 seconds (8 allocations: 11.258 MiB)

julia> @time sort!(4^7*repeat(1:30, 4^7); alg=CombSort);
  4.866824 seconds (4 allocations: 7.500 MiB)
```
---

Proof

Take an arbitrary `k`, let `m = 4k`, and let `n = m*4^7`. Consider the first 7 intervals for an input of length `n`: `[n*(3/4)^i for i in 1:7] == [m*4^7*(3/4)^i for i in 1:7] == [m*4^(7-i)*3^i for i in 1:7]`. Notice that each interval is divisible by `m`. 

Now, construct a pathological input `v = repeat(1:m, 4^7)`. This input has the property `v[i] == v[i+*jm]` for any intergers `i` and `j` which yield inbounds indices. Consequently, the first 7 passes cannot alter `v` at all.

_Informal interlude: There are still a lot of low numbers near the end of the list, and the remaining passes will have a hard time moving them to the beginning because their intervals are fairly small._

Consider the elements `1:k` that fall in the final quarter of `v`. There are `k*4^7/4 = n/16` such elements. Each of them must end up in the first quarter of the list once sorted, so they must each travel a total of at least `n/2` slots (in reality they must each travel more than this, but all I claim is a lower bound).

To recap, we have established `n/16` elements that must travel at least `n/2` slots, and that they do not travel at all in the first 7 passes. The remaining comb passes have intervals no greater than `[n*(3/4)^i for i in 8:inf]`. The furthest an elemental can move toward the start of the vector in a single pass is the interval size of that pass, so the furthest an element can move toward the start of the vector in all remaining passes combined is `sum([n*(3/4)^i for i in 8:inf]) = n*(3/4)^8 / (1 - 3/4) = 4n*(3/4)^8 < 0.401n`. Thus, after all the comb passes are compete, we will still have `n/16` elements that have to move at least `0.099n` slots toward the start of the vector. Insertion sort, which can only move one swap at a time will require `0.099n*n/4 > .024n^2` swaps to accomplish this. Therefore, the worst case runtime of this algorithm is `Ω(n^2)`.

It is structurally impossible for this algorithm to take more than `O(n^2)` time, so we can conclude `Θ(n^2)` is a tight asymptotic bound on the worst case runtime of this implementation of combsort. (A similar analysis holds for any geometric interval distribution).

---

We can verify the math in this proof empirically:

<details>

<summary> code </summary>

```julia
function comb!(v)
    lo, hi = extrema(eachindex(v))
    interval = (3 * (hi-lo+1)) >> 2
    while interval > 1
        for j in lo:hi-interval
            a, b = v[j], v[j+interval]
            v[j], v[j+interval] = b < a ? (b, a) : (a, b)
        end
        interval = (3 * interval) >> 2
    end
    v
end

function count_insertion_sort!(v)
    count = 0
    lo, hi = extrema(eachindex(v))
    for i = lo+1:hi
        j = i
        x = v[i]
        while j > lo && x < v[j-1]
            count += 1
            v[j] = v[j-1]
            j -= 1
        end
        v[j] = x
    end
    count
end

K = 1:6
M = 4 .* K
N = M .* 4^7
swaps = [count_insertion_sort!(comb!(repeat(1:m, 4^7))) for m in M]

using Plots
plot(N, swaps, label="actual swaps", xlabel="n", ylabel="swaps", legend=:topleft)
plot!(N, .024N.^2, label="theoretical minimum")
```

</details>

Results
<img width="598" alt="Screen Shot 2022-10-05 at 5 02 03 PM" src="https://user-images.githubusercontent.com/60898866/194047014-8a96fddb-60da-4f81-accc-26f05375221d.png">

---

The proof conveniently provides us with a pathological input to test. So, even more empirically, we can simply measure runtime.

<details>

<summary> Code </summary>

```julia
# multiply by a large number ot avoid dispatch to counting sort
make_vector(m) = 4^7*repeat(1:m, 4^7)
x = 1:20
n       = 4^7*x
comb    = [(x = make_vector(m); @elapsed(sort!(x; alg=CombSort))) for m in x]
default = [(x = make_vector(m); @elapsed(sort!(x              ))) for m in x]
theory  = .024n.^2 / 1.6e9 # 1.6 ghz clock speed

plot(n, comb, label="comb sort", xlabel="n", ylabel="time (s)", legend=:topleft)
plot!(n, default, label="default sort")
plot!(n, theory, label="theoretical minimum")
```

</details>

Results

<img width="592" alt="Screen Shot 2022-10-05 at 5 18 43 PM" src="https://user-images.githubusercontent.com/60898866/194048696-d4f0adc4-7ec7-4cbc-8238-2197618731c9.png">